File: ts/src/chaos.ts
/**
 * ChaosMeshLite: chaos experiments as CI-runnable SLO assertions.
 *
 * Inject deterministic faults (added latency, periodic errors) into a stream of
 * observed call latencies, then assert that Service Level Objectives still hold.
 * The fault schedule is deterministic (not RNG) so results are reproducible and
 * identical across every language port.
 */

/** Faults injected into an experiment. The defaults inject nothing. */
export class FaultPolicy {
  /** Latency added to every non-failed call. */
  readonly addedLatencyMs: number;
  /**
   * Fails calls whose 1-indexed position is a multiple of this value.
   * A value of 0 means no calls ever fail.
   */
  readonly failEveryN: number;

  constructor(addedLatencyMs = 0.0, failEveryN = 0) {
    this.addedLatencyMs = addedLatencyMs;
    this.failEveryN = failEveryN;
  }
}

/** The measured outcome of a single experiment run. */
export interface ExperimentResult {
  /** Number of observed calls, including failures. */
  readonly totalCalls: number;
  /** Number of injected failures. */
  readonly failures: number;
  /** successes/total rounded to six decimals (0 when there were no calls). */
  readonly successRate: number;
  /** 95th-percentile latency of the successful calls. */
  readonly p95LatencyMs: number;
}

/** A Service Level Objective to assert against an {@link ExperimentResult}. */
export interface Slo {
  /** Lowest acceptable success rate. */
  readonly minSuccessRate: number;
  /** Highest acceptable p95 latency. */
  readonly maxP95LatencyMs: number;
}

/** Whether an experiment met its SLO and, if not, why. */
export interface SloResult {
  /** True when there are no breaches. */
  readonly passed: boolean;
  /** Success rate of the evaluated result. */
  readonly successRate: number;
  /** p95 latency of the evaluated result. */
  readonly p95LatencyMs: number;
  /** Human-readable breach reasons (empty when passed). */
  readonly breaches: string[];
}

/**
 * Returns the nearest-rank percentile of `values` for `p` in [0, 1].
 * An empty array yields 0. The input array is not modified.
 */
export function percentile(values: number[], p: number): number {
  if (values.length === 0) {
    return 0.0;
  }
  const ordered = [...values].sort((a, b) => a - b);
  const rank = Math.ceil(p * ordered.length);
  const idx = Math.min(Math.max(rank - 1, 0), ordered.length - 1);
  return ordered[idx];
}

/**
 * Applies `policy` to `baseLatenciesMs` and measures the outcome. Failed calls
 * contribute no latency sample; every surviving call has `addedLatencyMs` added
 * to its base latency.
 */
export function runExperiment(baseLatenciesMs: number[], policy: FaultPolicy): ExperimentResult {
  const latencies: number[] = [];
  let failures = 0;
  for (let i = 0; i < baseLatenciesMs.length; i++) {
    const oneIndexed = i + 1;
    const isFailure = policy.failEveryN > 0 && oneIndexed % policy.failEveryN === 0;
    if (isFailure) {
      failures++;
      continue;
    }
    latencies.push(baseLatenciesMs[i] + policy.addedLatencyMs);
  }

  const total = baseLatenciesMs.length;
  const successes = total - failures;
  const successRate = total === 0 ? 0.0 : round6(successes / total);
  return {
    totalCalls: total,
    failures,
    successRate,
    p95LatencyMs: round6(percentile(latencies, 0.95)),
  };
}

/**
 * Checks `result` against `slo` and returns a pass/fail verdict with breach
 * reasons. The success-rate breach (if any) is reported before the latency
 * breach.
 */
export function evaluateSlo(result: ExperimentResult, slo: Slo): SloResult {
  const breaches: string[] = [];
  if (result.successRate < slo.minSuccessRate) {
    breaches.push(
      `success_rate ${result.successRate.toFixed(4)} < ${slo.minSuccessRate.toFixed(4)}`,
    );
  }
  if (result.p95LatencyMs > slo.maxP95LatencyMs) {
    breaches.push(
      `p95 ${result.p95LatencyMs.toFixed(2)}ms > ${slo.maxP95LatencyMs.toFixed(2)}ms`,
    );
  }
  return {
    passed: breaches.length === 0,
    successRate: result.successRate,
    p95LatencyMs: result.p95LatencyMs,
    breaches,
  };
}

function round6(v: number): number {
  return Math.round(v * 1_000_000.0) / 1_000_000.0;
}

File: go/chaos.go
// Package chaosmeshlite runs chaos experiments as CI-runnable SLO assertions.
//
// It injects deterministic faults (added latency, periodic errors) into a
// stream of observed call latencies, then asserts that Service Level
// Objectives still hold. The fault schedule is deterministic (not RNG) so
// results are reproducible and identical across every language port.
package chaosmeshlite

import (
	"fmt"
	"math"
	"slices"
)

// FaultPolicy describes the faults injected into an experiment. The zero value
// (no added latency, FailEveryN == 0) injects nothing.
type FaultPolicy struct {
	// AddedLatencyMs is added to every non-failed call's latency.
	AddedLatencyMs float64
	// FailEveryN fails calls whose 1-indexed position is a multiple of N.
	// A value of 0 means no calls ever fail.
	FailEveryN int
}

// ExperimentResult is the measured outcome of a single experiment run.
type ExperimentResult struct {
	// TotalCalls is the number of observed calls, including failures.
	TotalCalls int
	// Failures is the number of injected failures.
	Failures int
	// SuccessRate is successes/total, rounded to six decimals (0 when empty).
	SuccessRate float64
	// P95LatencyMs is the 95th-percentile latency of successful calls.
	P95LatencyMs float64
}

// Slo is a Service Level Objective to assert against an ExperimentResult.
type Slo struct {
	// MinSuccessRate is the lowest acceptable success rate.
	MinSuccessRate float64
	// MaxP95LatencyMs is the highest acceptable p95 latency.
	MaxP95LatencyMs float64
}

// SloResult reports whether an experiment met its SLO and why not.
type SloResult struct {
	// Passed is true when there are no breaches.
	Passed bool
	// SuccessRate mirrors the evaluated ExperimentResult.
	SuccessRate float64
	// P95LatencyMs mirrors the evaluated ExperimentResult.
	P95LatencyMs float64
	// Breaches holds human-readable reasons the SLO failed (empty when passed).
	Breaches []string
}

// Percentile returns the nearest-rank percentile of values for p in [0,1].
// An empty slice yields 0. The input slice is not modified.
func Percentile(values []float64, p float64) float64 {
	if len(values) == 0 {
		return 0.0
	}
	ordered := slices.Clone(values)
	slices.Sort(ordered)
	rank := int(math.Ceil(p * float64(len(ordered))))
	idx := min(max(rank-1, 0), len(ordered)-1)
	return ordered[idx]
}

// RunExperiment applies policy to baseLatenciesMs and measures the outcome.
// Failed calls contribute no latency sample; every surviving call has
// AddedLatencyMs added to its base latency.
func RunExperiment(baseLatenciesMs []float64, policy FaultPolicy) ExperimentResult {
	latencies := make([]float64, 0, len(baseLatenciesMs))
	failures := 0
	for i, base := range baseLatenciesMs {
		oneIndexed := i + 1
		isFailure := policy.FailEveryN > 0 && oneIndexed%policy.FailEveryN == 0
		if isFailure {
			failures++
			continue
		}
		latencies = append(latencies, base+policy.AddedLatencyMs)
	}

	total := len(baseLatenciesMs)
	successes := total - failures
	successRate := 0.0
	if total != 0 {
		successRate = round6(float64(successes) / float64(total))
	}
	return ExperimentResult{
		TotalCalls:   total,
		Failures:     failures,
		SuccessRate:  successRate,
		P95LatencyMs: round6(Percentile(latencies, 0.95)),
	}
}

// EvaluateSlo checks result against slo and returns a pass/fail verdict with
// breach reasons. The success-rate breach (if any) is reported before the
// latency breach.
func EvaluateSlo(result ExperimentResult, slo Slo) SloResult {
	breaches := []string{}
	if result.SuccessRate < slo.MinSuccessRate {
		breaches = append(breaches,
			fmt.Sprintf("success_rate %.4f < %.4f", result.SuccessRate, slo.MinSuccessRate))
	}
	if result.P95LatencyMs > slo.MaxP95LatencyMs {
		breaches = append(breaches,
			fmt.Sprintf("p95 %.2fms > %.2fms", result.P95LatencyMs, slo.MaxP95LatencyMs))
	}
	return SloResult{
		Passed:       len(breaches) == 0,
		SuccessRate:  result.SuccessRate,
		P95LatencyMs: result.P95LatencyMs,
		Breaches:     breaches,
	}
}

func round6(v float64) float64 {
	return math.Round(v*1_000_000.0) / 1_000_000.0
}
